data/blog/2022-08-01---10-things.md

```diff
@@ -0,0 +1,144 @@
+---
+title: "7 things I've learned building a modern TUI framework"
+---
+
+I've be working on [Textual](https://github.com/Textualize/textual) for over a year now. Here's a few things I've discovered (or re-discovered) regarding working with terminals in Python, and software development in general.
+
+&mdash; Will McGugan (CEO / Founder) [@willmcgugan](https://twitter.com/willmcgugan)
+
+### Terminals are fast
+
+A modern terminal emulator is a remarkably sophisticated piece of software. The protocol they run may be [ancient](https://en.wikipedia.org/wiki/Teleprinter) but likely powered by the same graphics technologies used in video games. And yet smooth animation is _not_ a given in the terminal. If you have ever tried any kind of visual effects in a terminal you may have been disappointed with flickering or _tearing_.
+
+But it is possible to achieve smooth animation as you can see from the following screencast. So what trick(s) are we pulling?
+
+<div style="text-align:center; margin: 1em 0;">
+    <iframe width="560" height="364" style="margin: 20px;" src="https://www.youtube.com/embed/k5m3CQT9yDM" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
+</div>
+
+There are a few factors which reduce flicker in the terminal. The first is the terminal emulator that you are using. Modern terminals tend to make better use of the GPU and will synchronize updates with your display to reduce flicker. Although in my experience there are other factors which have a greater impact on reducing flicker than your choice of emulator. Even on older terminals, you can generally get flicker-free animation with this _one little trick_ (actually there are three).
+
+The first _trick_ would be overwrite don't clear. If you clear the "screen" and then add new content, you risk seeing a blank or partially blank frame for a brief moment, leading to painful flicker. It's far better to overwrite the content in the terminal entirely so that there is no intermediate blank frame.
+
+The second _trick_ would be to write new content in a single write to standard output. It may be convenient to make multiple `file.write` calls to display an update, but as with the blank frame you risk a partial update becoming visible.
+
+The third _trick_ would be to use the [Synchronised Output](https://gist.github.com/christianparpart/d8a62cc1ab659194337d73e399004036) protocol; a relatively new addition to the terminal protocol, but already supported by many terminals. Details in the link, but the gist is that you tell the terminal when you begin and end a frame. It can then use this information to deliver flicker free updates.
+
+With these three tricks in place you can create very smooth animation as long as you can deliver updates at regular intervals. Textual uses 60fps as a baseline. Any more than that probably isn't going to be noticeable.
```

Comment:
With terminal stuff though?

Reply:
Don't see why it would make any difference if it's in the terminal if we're talking about animation. You notice it in the OS quite quickly - if you drag a window around (or even just move your cursor) it looks more fluid at 144hz vs 60hz. Doesn't matter much though, was just a note

Reply:
It's more that in order to perceive 120 / 144hz animation in the terminal, something would have to be moving at at least 120 cells per / second. I guess its not impossible, but it wouldn't happen that often. Unless you are building a terminal game maybe...

Reply:
Good point! Never thought about that. I guess 60 cells per second if you're using the "interpolation" trick we use e.g. for the tab animation. But yeah, seems unlikely :)